src/models/ListRule.ts

```diff
@@ -15,33 +15,140 @@ limitations under the License.
 */
 
 import { MatrixGlob } from "matrix-bot-sdk";
+import { EntityType } from "./PolicyList";
 
-export const RECOMMENDATION_BAN = "m.ban";
-export const RECOMMENDATION_BAN_TYPES = [RECOMMENDATION_BAN, "org.matrix.mjolnir.ban"];
+export enum Recommendation {
+    /// The rule recommends a "ban".
+    ///
+    /// The actual semantics for this "ban" may vary, e.g. room ban,
+    /// server ban, ignore user, etc. To determine the semantics for
+    /// this "ban", clients need to take into account the context for
+    /// the list, e.g. how the rule was imported.
+    Ban = "m.ban",
 
-export function recommendationToStable(recommendation: string, unstable = false): string|null {
-    if (RECOMMENDATION_BAN_TYPES.includes(recommendation)) return unstable ? RECOMMENDATION_BAN_TYPES[RECOMMENDATION_BAN_TYPES.length - 1] : RECOMMENDATION_BAN;
-    return null;
+    /// The rule specifies an "opinion", as a number in [-100, +100],
+    /// where +100 represents a user who is absolutely trusted and
+    /// -100 represents a user who is absolutely untrusted.
+    Opinion = "org.matrix.msc3845.opinion"
 }
 
-export class ListRule {
+/**
+ * All types for `m.ban`
+ */
+const RECOMMENDATION_BAN_VARIANTS = [
+    // Stable
+    Recommendation.Ban,
+    // Unstable prefix, for compatibility.
+    "org.matrix.mjolnir.ban"
+];
 
-    private glob: MatrixGlob;
+/**
+ * All types for `m.ban`
+ */
+const RECOMMENDATION_OPINION_VARIANTS: string[] = [
+    // Unstable
+    Recommendation.Opinion
+];
 
-    constructor(public readonly entity: string, private action: string, public readonly reason: string, public readonly kind: string) {
-        this.glob = new MatrixGlob(entity);
-    }
+export const OPINION_MIN = -100;
+export const OPINION_MAX = +100;
 
+// FIXME: This function is only ever called with a constant?
+export function recommendationToStable(recommendation: string): Recommendation | null {
+    if (RECOMMENDATION_BAN_VARIANTS.includes(recommendation)) return Recommendation.Ban;
+    if (RECOMMENDATION_OPINION_VARIANTS.includes(recommendation)) return Recommendation.Opinion;
+    return null;
+}
+
+/**
+ * Representation of a rule within a Policy List.
+ */
+export abstract class ListRule {
     /**
-     * The recommendation for this rule, or `null` if there is no recommendation or the recommendation is invalid.
-     * Recommendations are normalised to their stable types.
+     * A glob for `entity`.
      */
-    public get recommendation(): string|null {
-        if (RECOMMENDATION_BAN_TYPES.includes(this.action)) return RECOMMENDATION_BAN;
-        return null;
```

Comment:
What's the reason for changing this getter? You need to be really careful to check where this behaviour is being relied upon https://github.com/matrix-org/mjolnir/pull/336/files#diff-ebc449200d5204df932d4794476267f12b9759b8dfcbf34b7dc92361870b5f3cR155 is the only example (I think) and needs changing to ensure the kind is upgraded to stable.

Reply:
Well, for `ListRuleBan`, the `public readonly` property will return exactly the same thing as the getter, with simpler code.

For `ListRuleOpinion`, the `public readonly` property will return something different, but there is no helping it. I'll need to go through the entire codebase with a fine comb to find out exactly where the fact that all rules are `m.ban` is currently harcoded. I expect that this is going to take me some time, hence the current PR that only does part of the job.

Reply:
>I'll need to go through the entire codebase with a fine comb to find out exactly where the fact that all rules are m.ban is currently harcoded. I expect that this is going to take me some time, hence the current PR that only does part of the job.

Well yes but you can't do part of the job without introducing a bug so it doesn't seem acceptable, though searching this myself there is only one instance (the one linked) where there would be a bug.

Reply:
I've structured my refactoring to ensure that the behavior remains unchanged for the time being. There may be slight theoretical differences in ill-formed rules if someone happens to create an ill-formed `ListRuleOpinion`, but that shouldn't have any consequence.

Do we agree so far?

Reply:
I'm not sure that we're talking about the same thing, but so long as we only make rules with `List.parse` i don't think this will be a problem anymore.